stunnel/stunnel.ml

```diff
@@ -42,154 +41,155 @@ let timeoutidle = ref None
+        "/usr/sbin/stunnel4";
+        "/usr/sbin/stunnel";
+        "/usr/bin/stunnel4";
+        "/usr/bin/stunnel";
+      ] in
+      let rec choose l =
+        match l with
+        | [] -> raise Stunnel_binary_missing
+        | p::ps ->
+          try Unix.access p [Unix.X_OK]; p
+          with _ -> choose ps
+      in
+      let path = choose choices in
+      cached_stunnel_path := Some path
+    )
 
 let stunnel_path () =
-	if Opt.is_none !cached_stunnel_path then
-		init_stunnel_path ();
-	Opt.unbox !cached_stunnel_path
+  let open Xapi_stdext_monadic in
+  if Opt.is_none !cached_stunnel_path then
+    init_stunnel_path ();
+  Opt.unbox !cached_stunnel_path
 
 module Unsafe = struct
   (** These functions are not safe in a multithreaded program *)
 
   (* Low-level (unsafe) function which forks, runs a 'pre_exec' function and
-	 then executes some other binary. It makes sure to catch any exception thrown by
-	 exec* so that we don't end up with two ocaml processes. *)
+     	 then executes some other binary. It makes sure to catch any exception thrown by
+     	 exec* so that we don't end up with two ocaml processes. *)
   let fork_and_exec ?(pre_exec=fun () -> ()) ?env (cmdline: string list) = 
-	let args = Array.of_list cmdline in
-	let argv0 = List.hd cmdline in
-	let pid = Unix.fork () in
-	if pid = 0 then begin
+    let args = Array.of_list cmdline in
+    let argv0 = List.hd cmdline in
+    let pid = Unix.fork () in
+    if pid = 0 then begin
       try
-		pre_exec ();
-		  (* CA-18955: xapi now runs with priority -3. We then set his sons priority to 0. *) 
-		  ignore_int (Unix.nice (-(Unix.nice 0)));
-		  ignore_int (Unix.setsid ());
-		  match env with
-		  | None -> Unix.execv argv0 args
-		  | Some env -> Unix.execve argv0 args env
+        pre_exec ();
+        (* CA-18955: xapi now runs with priority -3. We then set his sons priority to 0. *) 
+        ignore_int (Unix.nice (-(Unix.nice 0)));
+        ignore_int (Unix.setsid ());
+        match env with
+        | None -> Unix.execv argv0 args
+        | Some env -> Unix.execve argv0 args env
       with _ -> exit 1
-	end else pid
-	  
-  (** File descriptor operations to be performed after a fork.
-      These are all safe in the presence of threads *)
-  type fd_operation = 
+    end else pid
+
+  (* File descriptor operations to be performed after a fork.
+   * These are all safe in the presence of threads *)
+  type fd_operation =
     | Dup2 of Unix.file_descr * Unix.file_descr
     | Close of Unix.file_descr
-		  
+
   let do_fd_operation = function
-	| Dup2(a, b) -> Unix.dup2 a b
-	| Close a -> Unix.close a
+    | Dup2(a, b) -> Unix.dup2 a b
+    | Close a -> Unix.close a
 end
 
-type pid = 
+type pid =
   | StdFork of int (** we forked and exec'ed. This is the pid *)
   | FEFork of Forkhelpers.pidty (** the forkhelpers module did it for us. *)
   | Nopid
 
-let string_of_pid = function
+(* let string_of_pid = function
```

Comment:
Dead code?

Reply:
Yes, left for reference. There is a lot that could be cleaned up or simplified in the stunnel code.